File: ejTasks/start_here/Authorization/Authorization.java
package Authorization;

import java.util.Set;

import Pii.Pii;

public class Authorization {
	private final Pii<String> m_username;
	private final Set<String> m_allowed;
	
	/*package*/ Authorization(Pii<String> a_username, Set<String> a_allowed) {
		m_username = a_username;
		m_allowed = a_allowed;
	}
	
	public Pii<String> getUsername() {
		return m_username;
	}
	
	public boolean allows(String a_right) {
		return m_allowed.contains(a_right);
	}
	
	public void throwIfNotAllowed(String a_right) throws InvalidAuth {
		if (!allows(a_right)) {
			throw new InvalidAuth(a_right);
		}
	}
}

File: sharpTasks/start_here/authorization/Authorization.cs
namespace AuthorizationNS
{
    public class Authorization
    {
        private readonly string m_username;
        private readonly ISet<string> m_allowed;

        public Authorization(string a_username, ISet<string> a_allowed)
        {
            m_username = a_username;
            m_allowed = a_allowed;
        }

        public string Username => m_username;

        public bool Allows(string a_right)
        {
            return m_allowed.Contains(a_right);
        }

        public void ThrowIfNotAllowed(string a_right)
        {
            if (!Allows(a_right))
            {
                throw new InvalidAuthException(a_right);
            }
        }
    }

    public class InvalidAuthException : Exception
    {
        private string m_right;

        public InvalidAuthException(string a_right) : base($"Unauthorized: {a_right}")
        {
            m_right = a_right;
        }

        public string Right => m_right;

        public string Explanation => "You do not have autherization to " + AuthMgr.INVALID_AUTH_EXPLANATIONS[m_right];
    }
}
